src/serac/numerics/functional/tests/functional_multiphysics.cpp

```diff
@@ -0,0 +1,166 @@
+// Copyright (c) 2019-2022, Lawrence Livermore National Security, LLC and
+// other Serac Project Developers. See the top-level LICENSE file for
+// details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+#include <fstream>
+#include <iostream>
+
+#include "mfem.hpp"
+
+#include <gtest/gtest.h>
+
+#include "axom/slic/core/SimpleLogger.hpp"
+#include "serac/infrastructure/input.hpp"
+#include "serac/serac_config.hpp"
+#include "serac/mesh/mesh_utils_base.hpp"
+#include "serac/numerics/expr_template_ops.hpp"
+#include "serac/numerics/stdfunction_operator.hpp"
+#include "serac/numerics/functional/functional.hpp"
+#include "serac/numerics/functional/tensor.hpp"
+
+using namespace serac;
+using namespace serac::profiling;
+
+template <typename T>
+void check_gradient(Functional<T>& f, mfem::Vector& U, mfem::Vector& dU_dt)
+{
+  int    seed    = 42;
+  double epsilon = 1.0e-8;
+
+  mfem::Vector dU(U.Size());
+  dU.Randomize(seed);
+
+  mfem::Vector ddU_dt(U.Size());
+  ddU_dt.Randomize(seed + 1);
+
+  auto U_plus = U;
+  U_plus.Add(epsilon, dU);
+
+  auto U_minus = U;
+  U_minus.Add(-epsilon, dU);
+
+  {
+    mfem::Vector df1 = f(U_plus, dU_dt);
+    df1 -= f(U_minus, dU_dt);
+    df1 /= (2 * epsilon);
+
+    auto [value, dfdU] = f(differentiate_wrt(U), dU_dt);
+    mfem::Vector df2   = dfdU(dU);
+
+    mfem::HypreParMatrix* dfdU_matrix = dfdU;
```

Comment:
I don't know what the best thing is to do here-- I don't really like the raw pointer. We could return a `std::unique_ptr< mfem::HypreParMatrix >`, or put an `mfem::HypreParMatrix` class member on the gradient object, and then it could just return a reference to that (this is what the QoI specialization is doing right now).

Other suggestions are welcome, too!

Reply:
I like the second option. This also means we could reuse the sparsity pattern during reassembly operations. When the upstreaming negotiations happen, we can discuss changing this if required.

Reply:
After trying to make it return `mfem::HypreParMatrix &` for a bit, it seems this doesn't work out so well after all (my bad). It seems like it makes more sense to really pass ownership of the matrix when assembling it, rather than returning a reference. I'm going to try implementing the interface below instead:

```cpp
struct Functional{
  ...
  struct Gradient{
    friend std::unique_ptr< mfem::HypreParMatrix > assemble(Gradient & grad);
  };
  ...
};
```

It would also be nice if the relevant mfem types (`mfem::HypreParMatrix`, `mfem::HypreParVector`) supported move constructors, so that we wouldn't need to wrap things in `std::unique_ptr`.